Lib/Src/adc_data.cpp

```diff
@@ -0,0 +1,95 @@
+/**
+ *@file adc_data.cpp
+ *@brief This files contains the methods for the CAdc class
+ *
+ */
+
+/*
+ *  Created on: Feb 14, 2022
+ *      Author: niuslar
+ */
+
+#include "adc_data.h"
+
+#define ADC_START_BIT   2
+#define ALARM_TOGGLE_MS 300
+
+CUartCom uart_adc(&huart2);
+
+CLog log_adc(&uart_adc, "ADC_data");
+
+/**
+ * @brief Calibrate and Start ADC with DMA
+ * @note The ADC will read all the channels once and then stop.
+ * To read the channels again, use the triggerADC() method.
+ * @param p_hadc Pointer to ADC handler
+ * @param adc_channels Number of active channels for the ADC
+ * 		(This depends on the hardware configuration)
+ */
+CAdc::CAdc(ADC_HandleTypeDef *p_hadc, uint8_t adc_channels)
+    : mp_hadc(p_hadc),
+      m_adc_channels(adc_channels)
+{
+    // Check mp_hadc is not null
+    if (!mp_hadc)
+    {
+        // Send error message
+        log_adc.log(CLog::LOG_ERROR, "Invalid pointer to ADC Handler");
+    }
+
+    // Allocate memory to store the incoming data
+    mp_adc_data_buf = new uint16_t[adc_channels];
```

Comment:
ouff. this is dangerous. runtime memory allocation error is very hard to catch. I suggest you statically allocate space for this data. 
You could do a fancy compile time gates.
```
// comment out if you want memory to be statically allocated.
#define DYNAMIC_MEMORY
#ifdef DYNAMIC_MEMORY
    mp_adc_data_buf = new uint16_t[adc_channels];
#elif
    mp_adc_data_buf = p_statically_allocated_memory;
#endif
```

Similarly you allocate the memory conditionally in .h file depending on the `#define` flag.

Reply:
I think using a ```#define ADC_CHANNELS 14``` in the header file is the best option. Easy to change and less prone to error.